### PR TITLE
Fix/enable UDP packet reassembly

### PR DESCRIPTION
**Pull request by szekelyisz**
UdpContext didn't care about pbuf chaining when receiving datagrams, leading
to fragments delivered to the application as individual packets.

### CODE REVIEW

**Comment by szekelyisz** [comment]
Thanks @d-a-v! Yeah I scratched my head a little because I found out in the meantime that reassembly is completely disabled in the LwIP 1.4 binary (although the feature is present in the source). Still I chose this approach for two reasons. First, having `#if`s around every reference of `pbuf->tot_len` or `pbuf->len` is ugly. Second, someone could choose to recompile LwIP 1.4 with reassembly support, and it would be impolite to leave them with the original bug intentionally.

I came across this bug when I wanted to send data for WS2812 LED strips encoded in OSC packets as blobs. The usual MTU of 1500 bytes would leave me with around maximum 470 LEDs per strip, so I was wondering if I could use IP-level fragmentation to squeeze more data into a single datagram and avoid application-level fragmentation/reassembly. After making sure that LwIP had this feature and worked correctly I realized the bug was in the Arduino code. Now I can receive data for thousands of LEDs.

File: libraries/ESP8266WiFi/src/include/UdpContext.h [hunk 1]
```diff
@@ -238,6 +238,8 @@ class UdpContext
         }
 
         auto deleteme = _rx_buf;
+
+        while(_rx_buf->len != _rx_buf->tot_len) _rx_buf = _rx_buf->next;
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/include/UdpContext.h]: while body on its own line please

File: libraries/ESP8266WiFi/src/include/UdpContext.h [hunk 2]
```diff
@@ -287,22 +289,26 @@ class UdpContext
...
             return 0;
 
-        size_t max_size = _rx_buf->len - _rx_buf_offset;
+        size_t max_size = _rx_buf->tot_len - _rx_buf_offset;
         size = (size < max_size) ? size : max_size;
-        DEBUGV(":urd %d, %d, %d\r\n", size, _rx_buf->len, _rx_buf_offset);
+        DEBUGV(":urd %d, %d, %d\r\n", size, _rx_buf->tot_len, _rx_buf_offset);
+
+        void* buf = pbuf_get_contiguous(_rx_buf, dst, size, size, _rx_buf_offset);
+        if(!buf) return 0;
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/include/UdpContext.h]: if body on its own line please

File: libraries/ESP8266WiFi/src/include/UdpContext.h [hunk 3]
```diff
@@ -287,22 +289,26 @@ class UdpContext
...
+
+        if(buf != dst) memcpy(dst, buf, size);
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/include/UdpContext.h]: memcpy on its own line please

File: libraries/ESP8266WiFi/src/include/UdpContext.h [hunk 4]
```diff
@@ -522,6 +528,90 @@ class UdpContext
...
+    /*
+     * Code in this conditional block is copied/backported verbatim from
+     * LwIP 2.1.2 to provide pbuf_get_contiguous.
+     */
+
+    static const struct pbuf *
+    pbuf_skip_const(const struct pbuf *in, u16_t in_offset, u16_t *out_offset)
+    {
+      u16_t offset_left = in_offset;
+      const struct pbuf *q = in;
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/include/UdpContext.h]: what is q? please either add a comment, or rename to something meaningful (e.g. it which is typical of iterator, or similar)

File: libraries/ESP8266WiFi/src/include/UdpContext.h [hunk 5]
```diff
@@ -522,6 +528,90 @@ class UdpContext
         reinterpret_cast<UdpContext*>(arg)->_recv(upcb, p, srcaddr, srcport);
     }
 
+#if LWIP_VERSION_MAJOR == 1
+    /*
+     * Code in this conditional block is copied/backported verbatim from
+     * LwIP 2.1.2 to provide pbuf_get_contiguous.
+     */
+
+    static const struct pbuf *
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/include/UdpContext.h]: It looks to me like this method doesn't have to be in the class definition. I suggest moving to the .cpp, maybe even move it out of the class altogether.
Of course, the other methods that make use of it should be moved as well.

File: libraries/ESP8266WiFi/src/include/UdpContext.h [hunk 6]
```diff
@@ -522,6 +528,90 @@ class UdpContext
...
+        q = q->next;
+      }
+      if (out_offset != NULL) {
+        *out_offset = offset_left;
+      }
+      return q;
+    }
+
+    u16_t
+    pbuf_copy_partial(const struct pbuf *buf, void *dataptr, u16_t len, u16_t offset)
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/include/UdpContext.h]: Does this have to be in the class definition? Can it be moved out into the cpp?

File: libraries/ESP8266WiFi/src/include/UdpContext.h [hunk 7]
```diff
@@ -522,6 +528,90 @@ class UdpContext
...
+          left = (u16_t)(left + buf_copy_len);
+          len = (u16_t)(len - buf_copy_len);
+          offset = 0;
+        }
+      }
+      return copied_total;
+    }
+
+    void *
+    pbuf_get_contiguous(const struct pbuf *p, void *buffer, size_t bufsize, u16_t len, u16_t offset)
```
Comment (devyte) [inline libraries/ESP8266WiFi/src/include/UdpContext.h]: Can this be moved to the cpp?

File: libraries/ESP8266WiFi/src/include/UdpContext.h (hunk 1 above)
Comment (szekelyisz) [reply libraries/ESP8266WiFi/src/include/UdpContext.h]: Fixed.

File: libraries/ESP8266WiFi/src/include/UdpContext.h (hunk 3 above)
Comment (szekelyisz) [reply libraries/ESP8266WiFi/src/include/UdpContext.h]: Fixed.

File: libraries/ESP8266WiFi/src/include/UdpContext.h (hunk 2 above)
Comment (szekelyisz) [reply libraries/ESP8266WiFi/src/include/UdpContext.h]: Fixed.

File: libraries/ESP8266WiFi/src/include/UdpContext.h (hunk 4 above)
Comment (szekelyisz) [reply libraries/ESP8266WiFi/src/include/UdpContext.h]: This code was copy-pasted from LwIP2 source, nevertheless I renamed it to`pbuf_it`.

File: libraries/ESP8266WiFi/src/include/UdpContext.h (hunk 5 above)
Comment (szekelyisz) [reply libraries/ESP8266WiFi/src/include/UdpContext.h]: There is no cpp. UdpContext is a header-only class. These 2 methods are used exclusively in this class, so I believe it's better to enclose it inside and not pollute the global namespace.

File: libraries/ESP8266WiFi/src/include/UdpContext.h (hunk 6 above)
Comment (szekelyisz) [reply libraries/ESP8266WiFi/src/include/UdpContext.h]: Again, same as above.

File: libraries/ESP8266WiFi/src/include/UdpContext.h (hunk 7 above)
Comment (szekelyisz) [reply libraries/ESP8266WiFi/src/include/UdpContext.h]: Again, same as above.

File: libraries/ESP8266WiFi/src/include/UdpContext.h (hunk 5 above)
Comment (devyte) [reply libraries/ESP8266WiFi/src/include/UdpContext.h]: I meant create one and move this in there. But there is additional cleanup to do here, so let's just leave this as is.

**Pull request merged**